### ISSUE TITLE
feat(automation): implement runtime triggers & add trigger listeners to orchestrator

### ISSUE THREAD

**Issue by ramfox** [qri-io/qri#1792]
### Test
```go
// orchestrator_test.go
package automation

// create test listeners w/ test triggers?

func TestListeners(t *testing.T) {
    // create bus
    // create handler that records when `event.ETTriggerWorkflow` is emitted
    // create orchestrator with 2 Listeners
    // emit triggers on bus
    // should have no `event.ETTriggerWorkflow` emitted
    // o.Start
    // emit triggers on bus
    // should have `event.ETTriggerWorkflow` emitted
    // o.Stop
    // emit triggers on bus
    // should not have more `event.ETTriggerWorkflow` events emitted
}
```

```go
// trigger_test.go
package trigger

// figure out how to create an AssertListenerSpec?

func TestCronListener(t *testing.T) {
    // create store
    // create bus
    // create handler that listens for `event.ETTriggerWorkflow`
    // create CronListener
    // ensure correct Type
    // ConstructTrigger from map, ensure expected
    // Start
    // Wait long enough to trigger
    // Stop
    // ensure we have emitted event
}

func TestCronTrigger(t *testing.T) {
    // create CronListener
    // great CronTrigger from `cl.ConstructTrigger` & options map
    // check enabled is expected
    // check ToMap creates map identical to options Map
    // Advance
    // check we have changed everything we need to for advancing
}
```

### declare Trigger interface & adjust workflow 
```go
package workflow

type Store interface {
    Updater
    Lister
}

type Updater interface {
    Create(did dataset.ID, pid peer.ID, triggers []trigger.Trigger) (*State, error)
}

type Lister interface {
    // List lists all the workflows in the store
    List(ctx context.Context, offset, limit int) ([]*Workflow, error)
    // ListEnabled lists all the enabled workflows in the store
    ListEnabled(ctx context.Context, offset limit int) ([]*Workflow, error)
}

type State struct {
    ...
    triggers []trigger.Trigger
}

func (m *MemStore) Create(did dataset.ID, pid peer.ID, triggers []trigger.Trigger) (*State, error)
func (m *MemStore) List(ctx context.Context, offset, limit int) ([]*Workflow, error)
func (m *MemStore) ListEnabled(ctx context.Context, offset, limit int) ([]*Workflow, error)
```
```go
// workflow/trigger.go
package workflow

// Type describes the type of trigger the system listens for
type TriggerType string

// Trigger has all the information needed to determine under what circumstances an `event.ETTriggerWorkflow`
// should be emitted. A Trigger should also know how to "advance" itself once a trigger was... triggered.
type Trigger interface {
    json.Marshaler
    json.Unmarshaler
    ID() string
    Enabled() boolean
    Type() TriggerType
    ToMap() map[string]interface{}
    Advance() error
}

type TriggerOptions struct {
    Type TriggerType
    Options map[string]interface{}
}
```

### define Listener interface & implement CronListener & CronTrigger
```go
// cron.go
package trigger

// Listener emits a `event.ETTriggerWorkflow` when a specific stimulus is triggered
// It knows how to start and stop itself, as well as how to create new triggers for its specific stimulus
type Listener interface {
    ConstructTrigger(m map[string]interface{}) (*workflow.Trigger, error)
    Type() workflow.TriggerType
    Start(ctx context.Context) error
    Stop() error
}

type CronListener struct { 
    pub event.Publisher
    store workflow.Lister
    interval time.Duration
}

func NewCronListener(store workflow.Lister, bus event.Bus) (*CronListener, error)
func (c *CronListener) ConstructTrigger(m map[string]interface{}) (*workflow.Trigger, error)
func (c *CronListener) Type() workflow.TriggerType { return CronTriggerType }
func (c *CronListener) Start(ctx context.Context) error {
    // start up loop to periodically check for triggers that has passed their run start times
    // basically just adapt existing Cron.Start function
}
func (c *CronListener) Stop() error

var _ Listener = (*CronListener)(nil)


const (
    CronTriggerType = workflow.TriggerType("CronTriggerType")
)

type CronTrigger struct {
    ID
    Type
    Enabled
    Start time.Time
    Perodicity iso8601.RepeatingInterval
    NextRunStart *time.Time
}
// much of this can be lifted from workflow/trigger.go
func (ct *CronTrigger) MarshalJSON() ([]byte, error)
func (ct *CronTrigger) UnmarshalJSON([]byte) error
func (ct *CronTrigger) ID() string
func (ct *CronTrigger) Enabled() boolean
func (ct *CronTrigger) Type() workflow.TriggerType
func (ct *CronTrigger) ToMap() map[string]interface{}
func (ct *CronTrigger) Advance() error

var _ workflow.Trigger = (*CronTrigger)(nil)
    
```

### adjust Orchestrator
```go
package automation

var (
    // ErrNoTriggerType indicates that no "type" information was given in the trigger option map
    ErrNoTriggerType = fmt.Errorf("")
    // ErrTriggerTypeNotFound indicates that the given trigger "type" given in the trigger option map
    // is not a type we support in the orchestrator
    ErrTriggerTypeNotFound = fmt.Errorf("")
)

type Orchestrator struct {
    ...
    listeners []*trigger.Listeners
}

func (o *Orchestrator) CreateWorkflow(did dataset.ID, pid peer.ID, triggerOpts []map[string]interface{}) (*workflow.State, error) {
    // iterate over trigger Opts
    triggers := []*workflow.Trigger
    for _, opt := range triggerOpts {
        type, ok := opt["type"]
        if !ok { return nil, ErrNoTriggerType }
        listener, err := o.listenerForTriggerType(type)
        if err != nil { return nil, err }    
        trigger, err := listener.ConstructTrigger(opt)
        triggers := append(triggers, trigger)
    }
    o.workflows.Create(did, pid, triggers)    
}

func (o *Orchestrator) listenerForTriggerType(type string) (*trigger.Listener, error) {
    // iterates over listeners for if they match the type
}

func (o *Orchestrator) StartListeners(ctx context.Context) error {
    // iterate over listeners & call listener.Start(ctx)
}

func (o *Orchestrator) StopListeners() error {
    // iterate over listeners & call listener.Stop(), only exiting when the listeners have completed stopping
}

func (o *Orchestrator) Shutdown() error {
    // call o.StopListeners()
}
```

### adjust orchestrator options
```go
type Options struct {
    ...
    listeners []*trigger.Listener
}

func OptListeners(l []*trigger.Listener) Option {
    return func(opt *Options) error {
        o.listeners = l
        return nil
    }
}
```

**Comment by ramfox** [comment]
Add `TriggerSource` interface to `trigger` & make sure `workflow.Store` satisfies it ala this comment:
https://github.com/qri-io/qri/pull/1797#issuecomment-850549868

**Comment by ramfox** [comment]
okay, been over here having a bit of a hard time working through something.

Previously, I was using `CronListener` as a model when mentally stepping through functionality. 
1) we spin up an orchestrator, that orchestrator spins up any configured listeners.
2) the listeners "Start" listening. For the `CronListener`, that means starting a loop that, for every given interval, fetches a list of TriggerSources from a TriggerStore. Since it is fetching that list each time, it is always getting an up-to-date list of triggers.
3) the cron listener iterates through the triggers and emits a `ETWorkflowTrigger` for any that satisfy

However, CronTriggers are unique because they are entirely self contained. The only outside force they rely on is time, so it is logical to use time to poll the TriggerStore.

However, when I started to code up the `RuntimeListener`, the listener sister to the `RuntimeTrigger`, I realized that we don't have the same guarantees with the `RuntimeListener` than we do with a `CronListener`. 

A `RuntimeListener` has a channel of strings, each time a string is sent over the channel, a `ETWorkflowTrigger` event is emitted with the string as the payload (this string is actually the associated workflow ID). NEVERMIND WAIT as long as someone has the channel & sends the workflow ID on it, the `RuntimeListener` never actually has to have any knowledge of the triggers its listening for, the runtime listener will just react when the string comes across the channel. And this is also okay for something like a webhook listener. If something triggers the `WebhookListener`, we can then query the `TriggerStore` to get any security information to ensure the webhook was triggered in good faith. We don't need to have any "active" listening happening.'

PHEW
